// ionic2-jpush/src/plugin-jpush.service.ts
/**
 * Created by youyou on 16/11/23.
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Rx';

declare var plugins:any;
declare var document:any;

@Injectable()
export class JPushService {

  private jPushPlugin = (<any>window).plugins ? (<any>window).plugins.jPushPlugin || null : null;
  warnText:string = '没有找到 jPushPlugin 对象 \n也许你是在浏览器环境下运行或者没有正确安装插件； \n如果没有在Platform 的 ready 方法中调用，也会出现这样的情况。\n了解：http://ionicframework.com/docs/v2/api/platform/Platform/';
  constructor () {

  }

  wrapEventObservable(event: string): Observable<any> {
    return new Observable( (observer:any) => {
      document.addEventListener(event, observer.next.bind(observer), false);
      return () => document.removeEventListener(event, observer.next.bind(observer), false);
    });
  }

  initJPushPlugin() {
    this.jPushPlugin = (<any>window).plugins ? (<any>window).plugins.jPushPlugin || null : null;
  }

  init() {
    this.initJPushPlugin();
    return new Promise((resolve,reject) =>{

      if(this.jPushPlugin){
        this.jPushPlugin.init();
        resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })
  }

  getRegistrationID(){
    this.initJPushPlugin();

    return new Promise((resolve,reject) =>{

      if(this.jPushPlugin){
        this.jPushPlugin.getRegistrationID( (id:any) => {
          if(id){
            resolve(id)
          }else{
            reject('获取ID失败')
          }
        } )
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }


    })
  }

  stopPush(){
    this.initJPushPlugin();

    return new Promise((resolve,reject) =>{

      if(this.jPushPlugin){
        this.jPushPlugin.stopPush();
        resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })
  }

  resumePush(){
    this.initJPushPlugin();
    return new Promise((resolve,reject) =>{
      if(this.jPushPlugin){
        this.jPushPlugin.resumePush();

        resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })
  }

  isPushStopped(){
    this.initJPushPlugin();
    return new Promise((resolve,reject) =>{

      if(this.jPushPlugin){
        this.jPushPlugin.isPushStopped( (result:any) => {
          if(result === 0 ){
            resolve(true)
          }else{
            reject(false)
          }
        } )
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }


    })
  }

  /**
   *
   * @param tags
   * @param alias
   * @returns {Promise<T>}
   */
  setTagsWithAlias(tags:Array<any>,alias:string){
    this.initJPushPlugin();
    return new Promise((resolve,reject) =>{
      if(this.jPushPlugin){
        this.jPushPlugin.setTagsWithAlias(tags,alias);
        resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })
  }

  /**
   *
   * @param tags
   * @returns {Promise<T>}
   */
  setTags(tags:Array<any>){
    this.initJPushPlugin();
    return new Promise((resolve,reject) =>{
      if(this.jPushPlugin){
        this.jPushPlugin.setTags(tags);
        resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })
  }

  /**
   *
   * @param alias
   * @returns {Promise<T>}
   */
  setAlias(alias:string){
    this.initJPushPlugin();

    return new Promise((resolve,reject) =>{
      if(this.jPushPlugin){
        this.jPushPlugin.setAlias(alias);
        resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })
  }

  /**
   *
   * @param value
   * @returns {Promise<T>}
   */
  setBadge(value:number){
    this.initJPushPlugin();
    return new Promise((resolve,reject) =>{
      if(this.jPushPlugin){
          this.jPushPlugin.setBadge(value);
          resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })

  }

  reSetBadge(){
    this.initJPushPlugin();

  return new Promise((resolve,reject) =>{
    if(this.jPushPlugin){
        this.jPushPlugin.reSetBadge();
        resolve('ok')
    }else {
      console.warn(this.warnText);
      reject('没有找到 jPushPlugin');
    }
  })

}

  /**
   *
   * @param value
   * @returns {Promise<T>}
   */
  setApplicationIconBadgeNumber(value:number) {
    this.initJPushPlugin();
    return new Promise((resolve, reject) => {
      if (this.jPushPlugin) {
          this.jPushPlugin.setApplicationIconBadgeNumber(value);
          resolve('ok')

      } else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })
  }

  getApplicationIconBadgeNumber() {
    this.initJPushPlugin();

    return new Promise((resolve,reject) =>{

      if(this.jPushPlugin){
          this.jPushPlugin.getApplicationIconBadgeNumber( (num:any) => {
            resolve(num)
          } )
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }


    })
  }

  /**
   *
   * @param id
   * @returns {Promise<T>}
   */
  clearNotificationById(id:number){
    this.initJPushPlugin();
    return new Promise((resolve,reject) =>{
      if(this.jPushPlugin){
          this.jPushPlugin.clearNotificationById(id);
          resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })

  }

  clearAllNotification(){
    this.initJPushPlugin();

    return new Promise((resolve,reject) =>{
      if(this.jPushPlugin){
          this.jPushPlugin.clearAllNotification();
          resolve('ok')
      }else {
        console.warn(this.warnText);
        reject('没有找到 jPushPlugin');
      }
    })

  }

  /**
   *
   * @returns {Observable<any>}
   */
  openNotification(){
    return this.wrapEventObservable('jpush.openNotification');
  }

  /**
   *
   * @returns {Observable<any>}
   */
  receiveNotification(){
    return this.wrapEventObservable('jpush.receiveNotification');
  }

  /**
   *
   * @returns {Observable<any>}
   */
  receiveMessage(){
    return this.wrapEventObservable('jpush.receiveMessage');
  }

  backgroundNotification(){
    return this.wrapEventObservable('jpush.backgroundNotification');
  }
}
