// ionic2-jpush/dist/plugin-jpush.service.js
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Rx';
export var JPushService = (function () {
    function JPushService() {
        this.jPushPlugin = window.plugins ? window.plugins.jPushPlugin || null : null;
        this.warnText = '没有找到 jPushPlugin 对象 \n也许你是在浏览器环境下运行或者没有正确安装插件； \n如果没有在Platform 的 ready 方法中调用，也会出现这样的情况。\n了解：http://ionicframework.com/docs/v2/api/platform/Platform/';
    }
    JPushService.prototype.wrapEventObservable = function (event) {
        return new Observable(function (observer) {
            document.addEventListener(event, observer.next.bind(observer), false);
            return function () { return document.removeEventListener(event, observer.next.bind(observer), false); };
        });
    };
    JPushService.prototype.initJPushPlugin = function () {
        this.jPushPlugin = window.plugins ? window.plugins.jPushPlugin || null : null;
    };
    JPushService.prototype.init = function () {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.init();
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    JPushService.prototype.getRegistrationID = function () {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.getRegistrationID(function (id) {
                    if (id) {
                        resolve(id);
                    }
                    else {
                        reject('获取ID失败');
                    }
                });
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    JPushService.prototype.stopPush = function () {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.stopPush();
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    JPushService.prototype.resumePush = function () {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.resumePush();
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    JPushService.prototype.isPushStopped = function () {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.isPushStopped(function (result) {
                    if (result === 0) {
                        resolve(true);
                    }
                    else {
                        reject(false);
                    }
                });
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    /**
     *
     * @param tags
     * @param alias
     * @returns {Promise<T>}
     */
    JPushService.prototype.setTagsWithAlias = function (tags, alias) {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.setTagsWithAlias(tags, alias);
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    /**
     *
     * @param tags
     * @returns {Promise<T>}
     */
    JPushService.prototype.setTags = function (tags) {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.setTags(tags);
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    /**
     *
     * @param alias
     * @returns {Promise<T>}
     */
    JPushService.prototype.setAlias = function (alias) {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.setAlias(alias);
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    /**
     *
     * @param value
     * @returns {Promise<T>}
     */
    JPushService.prototype.setBadge = function (value) {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.setBadge(value);
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    JPushService.prototype.reSetBadge = function () {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.reSetBadge();
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    /**
     *
     * @param value
     * @returns {Promise<T>}
     */
    JPushService.prototype.setApplicationIconBadgeNumber = function (value) {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.setApplicationIconBadgeNumber(value);
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    JPushService.prototype.getApplicationIconBadgeNumber = function () {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.getApplicationIconBadgeNumber(function (num) {
                    resolve(num);
                });
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    /**
     *
     * @param id
     * @returns {Promise<T>}
     */
    JPushService.prototype.clearNotificationById = function (id) {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.clearNotificationById(id);
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    JPushService.prototype.clearAllNotification = function () {
        var _this = this;
        this.initJPushPlugin();
        return new Promise(function (resolve, reject) {
            if (_this.jPushPlugin) {
                _this.jPushPlugin.clearAllNotification();
                resolve('ok');
            }
            else {
                console.warn(_this.warnText);
                reject('没有找到 jPushPlugin');
            }
        });
    };
    /**
     *
     * @returns {Observable<any>}
     */
    JPushService.prototype.openNotification = function () {
        return this.wrapEventObservable('jpush.openNotification');
    };
    /**
     *
     * @returns {Observable<any>}
     */
    JPushService.prototype.receiveNotification = function () {
        return this.wrapEventObservable('jpush.receiveNotification');
    };
    /**
     *
     * @returns {Observable<any>}
     */
    JPushService.prototype.receiveMessage = function () {
        return this.wrapEventObservable('jpush.receiveMessage');
    };
    JPushService.prototype.backgroundNotification = function () {
        return this.wrapEventObservable('jpush.backgroundNotification');
    };
    JPushService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    JPushService.ctorParameters = [];
    return JPushService;
}());
